### ISSUE TITLE
YOLO v3 training on stock exchanges graphics for pattern recognition

### ISSUE THREAD

**Issue by xinger** [AlexeyAB/darknet#1538]
Hi, first of all, thank you for detailed readme and job that you doing with this repo.

I have a task for visual recognition of stock exchange figures (f.e. head and shoulders, not a crypto btw), and this task is so hard to resolve by math (for me), so, I decide to try YOLO v3.

I managed to prepare a huge dataset (which continue to grown) with almost ±4000 of pictures. The dataset contains 11 patterns, every pattern contains 2 classes — first class will say about the buying process, the second one about the selling process — so, classes are mirrored. In total, I have 22 classes. 

I did a lot of experiments with different kinds of the dataset and .cfg file:
I have begun from 1 class for test training and ended up with the full dataset for training with 4k pictures for tests. But, after 50k iterations I still have no any even close to satisfactory results. 

For example, in this image:

![gizxe6za](https://user-images.githubusercontent.com/3993459/45049657-ed1d6400-b087-11e8-95a3-b748be6c7ef1.png)
Yolo-detect recognized a photo part and whole graphic part, not patterns inside the graphic.

If I’ll run the same command with -thresh 0, this will be a result:

![q8eonufg](https://user-images.githubusercontent.com/3993459/45049686-0aeac900-b088-11e8-96e1-c910fff08e6e.png)

So, classes exist in weight files. I also checked every 5k iteration weight file, and still no good looking result, I don’t think it could be an “over-training” problem.

Now, few details about the dataset itself:
Patterns are manually prepared and stored in DB, so I can export them with any visual style or data format. 

I decided, that for the neural network it will be easier if I will prepare the dataset in style, where only 4 colors and every candle is 1-pixel width. For example, on this picture, in the middle you can see pattern itself:

![zlwljlzg](https://user-images.githubusercontent.com/3993459/45049728-2b1a8800-b088-11e8-8d5f-a60fe6727898.png)

This is the example of «pure» pattern, without any marked data on it, as you see, they are looking pretty close to each other:
![1lnjctrw](https://user-images.githubusercontent.com/3993459/45049780-5b622680-b088-11e8-9b66-797c182780c9.png)
Here is my final [config file](https://www.dropbox.com/s/ly2uhfg2q4kcusw/yolov3-22-class-v7.cfg?dl=0)
As you wrote at your _readme.md_, I also changed strings in .cfg about layers = -1, 11 and stride=4, because of patterns, they can be pretty small.

**Why, I think, it isn’t working for me:**
1) I run training on the base weight file darknet53.conv.74, that is could be strange because this weight was prepared for photos?

2) I think I need dramatically change my .cfg file because it was prepared for “photo recognition”, not “graphic patterns recognition”

3) Maybe, something wrong in pictures style that I feeding to YOLO?

Could you please share any thoughts about my issue? 
I’m desperate and looking for any advice 🙂

**Comment by PeterQuinn925** [comment]
Interesting idea. My first thought was that you should smooth the curves and use math to find the patterns you're looking for. I'm not sure this is the best tool for the job, but there's no reason that I can think of that your approach couldn't work. Some things to think about.

Do the colors matter? 
Does the scale matter? Yolo scales the images.
Does rotation matter? 
You don't want to use any of the precomputed weights since your data is completely different. If you omit the weight file from your training it will start with random weights, which is what you'll want to do.

**Comment by AlexeyAB** [comment]
@xinger Hi,

* Can you show several screenshots of your labeled training images in the  https://github.com/AlexeyAB/Yolo_mark ?

* Can you show screenshot of such command? 
    `./darknet detector calc_anchors data/obj.data -num_of_clusters 9 -width 416 -height 416 -show`

* What mAP can you get on Training and Validation datasets?

----

Also:

1. Yes, may be you should train without pre-trained weights (without `darknet53.conv.74`)

2. What is the min, max, average resolution of your Training/Test images?

3. I think that the main problem is - broken this rule:  https://github.com/AlexeyAB/darknet#how-to-improve-object-detection

> General rule - your training dataset should include such a set of relative sizes of objects that you want to detect:
> 
> * `train_network_width * train_obj_width / train_image_width ~= detection_network_width * detection_obj_width / detection_image_width`
> * `train_network_height * train_obj_height / train_image_height ~= detection_network_height * detection_obj_height / detection_image_height`

I.e. if you use the same network size for Training and Detection, then should be:
`train_obj_width / train_image_width ~= detection_obj_width / detection_image_width`

----

Your cfg-file looks good.

> If I’ll run the same command with -thresh 0, this will be a result:

So it should always be.

> I have a task for visual recognition of stock exchange figures (f.e. head and shoulders, not a crypto btw)

Yes, may be it would not work for the crypto, since there is a very manipulative market.

----

So I think the main problem is that for Training you use images with relative size of pattern =~ 1.0, but for Detection you use images with relative size of pattern =~ 0.1.
Any modern Neural Network isn't Scale-invariant, so you should train with the same relative size, that you want to detect.

**Comment by xinger** [comment]
@PeterQuinn925 

I've tried the method which was described in this [article](https://systematicinvestor.wordpress.com/2012/05/22/classical-technical-patterns/) and showed the result to our team trader, and he's saying , that he wasn't impressed by the result. Basically the result is like [ZigZag indicator](https://www.investopedia.com/terms/z/zig_zag_indicator.asp), and actually using its tops and bottoms we could search patterns by a simple algorithm, but unfortunately it does not fit. You need to empirically select the settings for the indicator for the every moment of the time and for the each currency.

About the colours. In the beginning, it seemed that colours have the sense, but as far we re going, the stronger is the feeling that the colours don't have the big sense. In fact, the form, which appears on the graphic is important by itself. 

The rotation doesn't need to be considered. There's only 180° rotation, which means, that the figure speaks about either the sale or purchase. But for that I separate different classes, and in dataset there are both these and those images.

I didn't know that u can not to specify the basic weights. Thank you!

**Comment by xinger** [comment]
@AlexeyAB 

> Can you show several screenshots of your labeled training images in the

![photo_2018-09-06_23-47-04](https://user-images.githubusercontent.com/3993459/45184308-47a4f480-b22f-11e8-8993-86247764d611.jpg)
![photo_2018-09-06_23-47-09](https://user-images.githubusercontent.com/3993459/45184314-4a074e80-b22f-11e8-9636-001627d903a1.jpg)
![photo_2018-09-06_23-47-10](https://user-images.githubusercontent.com/3993459/45184317-4bd11200-b22f-11e8-986f-b11e8b5f1c4e.jpg)
![photo_2018-09-06_23-47-12](https://user-images.githubusercontent.com/3993459/45184320-4d023f00-b22f-11e8-812a-bdead1c2458d.jpg)
![photo_2018-09-06_23-47-14](https://user-images.githubusercontent.com/3993459/45184327-4f649900-b22f-11e8-8f7c-bb0f80800fe7.jpg)

> Can you show screenshot of such command?

![photo_2018-09-06_23-47-31](https://user-images.githubusercontent.com/3993459/45184378-702cee80-b22f-11e8-8c7d-dd7c066685c7.jpg)

> What mAP can you get on Training and Validation datasets?

![photo_2018-09-06_23-47-22](https://user-images.githubusercontent.com/3993459/45184398-7d49dd80-b22f-11e8-8149-829577ecaf65.jpg)

> Yes, may be you should train without pre-trained weights (without darknet53.conv.74)

How can I start training without any pre-trained weights?

> I think that the main problem is - broken this rule: 

For example, I have width=160 and height=160 in train .cfg file, and my dataset png files are 160x160, should I change it somehow?

> So I think the main problem is that for Training you use images with relative size of pattern =~ 1.0, but for Detection you use images with relative size of pattern =~ 0.1.

I will try to use 160 x 160 for detection too, thank you for your answer and this idea 🙂

**Comment by AlexeyAB** [comment]
Also may be you should add paramter `flip=0` in the `[net]` section in your cfg-file. Otherwise data augmentation function will do horizontal-flip (rotate the image horizontally around vertical line).

**Comment by xinger** [comment]
> 1. Do you use width=160 height=160 for Training and Detection?...

For all questions answer is "Yes" because now, I tried to detect any patterns on 160x160 png, and I have a result!

![screenshot 2018-09-09 17 35 56](https://user-images.githubusercontent.com/3993459/45265641-d3e53080-b456-11e8-8352-8073c34ec36f.png)


Now, I have a small question — on this gif, you can see that the detected pattern is disappearing from a few frames, why could it happen?

![1](https://user-images.githubusercontent.com/3993459/45265599-296d0d80-b456-11e8-97a8-da8f465ced91.gif)

But here is everything okay with tracking:

![2](https://user-images.githubusercontent.com/3993459/45265607-37229300-b456-11e8-8326-e4b2b8ed88a9.gif)

And last question — classes in the .cfg file, they are starting from 0? Or from that number, that I specify in every .txt for labeled data?

**Comment by AlexeyAB** [comment]
In the cfg-file `classes=` are starting from 1, because you can't train for 0 classes :)
In labels txt-files `<class_id>` is starting from 0: https://github.com/AlexeyAB/darknet#how-to-train-to-detect-your-custom-objects

> Where:
> 
> <object-class> - integer object number from 0 to (classes-1)

----

> Now, I have a small question — on this gif, you can see that the detected pattern is disappearing from a few frames, why could it happen?
> 
> ![1](https://user-images.githubusercontent.com/3993459/45265599-296d0d80-b456-11e8-97a8-da8f465ced91.gif)

Perhaps some price movement before and after the pattern also affect the result, in artificial intelligence opinion )

Yolo takes in account whole image when you train it. The more priority for that is inside of truth_bounded_box, the less priority for that is outside of truth_bounded_box.
I.e. it is looking for any correlations and dependencies on the image, i.e. it is looking at prices before your pattern and after your pattern too.

To avoid this, you need to increase the training sample, up to 2000 images per class.

You should have about 42 000 images for 21 classes.

----

If you do not want to rely on the opinion of artificial intelligence, then you can color the white movement of the price before and after the pattern, on the training sample.

**Comment by AlexeyAB** [comment]
@xinger 

What mAP can you get currently on Training and Test dataset?

----

> Does a strong scatter in image quantity influence per class on detection quality? For example, the first class has 2000 images, and the second 400? The second will just be recognised worse? Or does it also has some influence on the first class?

In this case: the class with 2000 images will be detected better than class with 400 images.
It is recommended to increase number of images for the 2nd class.

----

> For increase the training dataset we came up to do 3 images on each pattern with a shift. Insofar as we have automatic chart scale, then with each shift there is a random vertical figure stretching.
> 
> Does it have any sense and don’t harm all in all?

It is not bad, it will allow to detect your pattern in different y-scales. It is good.

----

> Also to increase the training set we do a vertical flip of the images. For example, if u ll vertically flip pattern in BUY, u ll get the same pattern but in SELL. In this way we increase the set in two times. How bad is it?

You should be sure that vertical flipped BUY pattern behaves the same way as SELL pattern. I'm not sure that this is so.

----

> We have tried to train without context, and the weights completely ‘refused to work’. We gave the images of the same size 160x160 and, as u recommended, all the candles before and after pattern have done white:
> and detect results is:
> Maybe something needs to be changed in config?

Try to mix: all images with context + 10% images without context (context painted white).
Then check mAP of weights that trained on two datasets: 
* all images with context 
* all images with context + 20% images without context

----

> How do u think, if there any sense to increase image resolution to 320x320? Will it give any growth of recognition accuracy?

I think no, since all your Training and Test images are 160x160.


----

> Which approach is most effective – one weights educated on 20 classes, or 20 weights educated on 1 class?

I think one weights file that is trained on 20 classes will be better. 
And if you have already solved the problem with correct dataset, then it is better to increase 2x-4x value of `filters=` in each [convolutional] layer in your cfg-file instead of using 20 weights files.

----


> Which weights testing methods are mostly effective? After all u have to test every 5k of weights on one training and plus u have to be able to compare different trainings between each other. How to find the optimal threshold for the specific weights?

What do you mean?
Just calculate accuracy (mAP) for each 10000 iterations by using this command: https://github.com/AlexeyAB/darknet#when-should-i-stop-training

`./darknet detector map data/obj.data yolo-obj.cfg backup\yolo-obj_10000.weights`

Then calculate mAP for weights-files which near with the best weights-file. For example, the best weights `yolo_obj_30000.weights` (85% mAP), then calculate mAP for `yolo_obj_29000.weights`, `yolo_obj_31000.weights`, ...

Once you found the weights-file with the highest mAP, then run this command with different `-thresh 0.2` , `-thresh 0.3` parameters, so you can found threshold parameter with indicators that you want: TP (true-positive), FP (false-positive), FN (false-negative).

If you don't know what does it mean, then just find `-thresh` for which Precision and Recall will be the ~same. And use this threshold for Detection.

**Comment by xinger** [comment]
Hi, @AlexeyAB! A lot of time has passed since our last communication, but we finally have some progress now 🙂

We wrote our own weight testing system, which rely on our specific task and based on the [Matthews correlation coefficient](https://en.wikipedia.org/wiki/Matthews_correlation_coefficient). We can now move a folder with weights to the testing system and it will test each weight in all available thresholds, then build a graph and give us one single ideal weight with a specific threshold that we will use. The testing system also compares different training results between each other.

Also, when we exporting a training set, we run an automatic test to identify various errors, for example, that the width of the pattern is less than the picture itself, or the fact that the coordinates of the pattern do not go beyond the boundaries of the picture, and so on. So, the dataset itself is checked every time before training.

We did a lot of trainings and experiments with different configs and found final .cfg that is suitable for us. https://www.dropbox.com/s/ixzndtnblipvbn6/v29.cfg?dl=0

Thank for your help before, it really helped us a lot. Could you please help us again? Because looks like we stuck a little bit.

All the patterns in the pictures from the training set are "stick" to the right edge of the picture on purpose. This is 100% true and tested on the system for testing the datasets, about which I wrote above. And we expect, that detections will also be sticked to the right.

But, after detection, the coordinates of some recognised boxes are "moving". Periodically, it is literally plus or minus a couple of pixels, which understandable and connected to a certain bias in detection, but sometimes we have very strange cases when the coordinates of the recognized boxes are moved plus or minus 10 px, 20 px or 30 px, this is despite the fact the total width of the image 160px.

![anomaly](https://user-images.githubusercontent.com/3993459/56078729-2d01b200-5df4-11e9-9f36-b6c91d7c36ea.png)

In this screenshot below you can see how often "shift" in the detection happens. The key is y2 coordinate and value is detection count, so 159 is exactly the right border. Anything over 159 means that the coordinates of the detection box located "outside" the area of the picture.

![shifts](https://user-images.githubusercontent.com/3993459/56078743-4efb3480-5df4-11e9-805f-c566295ea2d3.jpg)

Could you say, how can we fix it?

**Comment by AlexeyAB** [comment]
@xinger Hi,

So network think that the pattern is there.

* if the most of detected objects with shifted coords are False Detection (FP), then just reject such detections if(shift > some_pixels)

* if the most of detected objects with shifted coords are True Detection (TP), then just force change the coordinate to required in your code
